src/pallene/pallenelib.lua

```diff
@@ -43,9 +43,73 @@ return [==[
 #include <stdarg.h>
 #include <locale.h>
 #include <math.h>
+#include <stdbool.h>
 
 #define PALLENE_UNREACHABLE __builtin_unreachable()
 
+/* Part of Pallene Tracer. */
+#define PALLENE_C_FRAMEENTER(L, name)   pt_fn_details_t _details = {                             \
+                                            .fn_name  = name,                                    \
+                                            .mod_name = PALLENE_SOURCE_FILE                      \
+                                        };                                                       \
+                                        pt_frame_t _frame = {                                    \
+                                            .type = PALLENE_TRACER_FRAME_TYPE_C,                 \
+                                            .shared = {                                          \
+                                                .details = &_details                             \
+                                            }                                                    \
+                                        };                                                       \
+                                        pallene_tracer_frameenter(L, &_frame)
+
+#define PALLENE_LUA_FRAMEENTER(L, sig)  pt_frame_t _frame = {                                    \
+                                            .type = PALLENE_TRACER_FRAME_TYPE_LUA,               \
+                                            .shared = {                                          \
+                                                .frame_sig = sig                                 \
+                                            }                                                    \
+                                        };                                                       \
+                                        pallene_tracer_frameenter(L, &_frame)
+
+#define PALLENE_GLOBAL_SETLINE(L, line) pallene_tracer_global_setline(L, line)
+#define PALLENE_SETLINE(line)           pallene_tracer_setline(&_frame, line)
+#define PALLENE_FRAMEEXIT(L, ...)       pallene_tracer_frameexit(L);                             \
+                                        return __VA_ARGS__
+
+
+/* PALLENE TRACER RELATED DATA-STRUCTURES. */
+
+/* Whether the frame is a Pallene->Pallene or Lua->Pallene call. */
+typedef enum frame_type {
+    PALLENE_TRACER_FRAME_TYPE_C,
+    PALLENE_TRACER_FRAME_TYPE_LUA
+} frame_type_t;
+
+/* Details of a single function such as what is the name
+   and where it is from. */
+typedef struct pt_fn_details {
+    const char *const fn_name;
+    const char *const mod_name;
+} pt_fn_details_t;
+
+/* A single frame representation. */
+typedef struct pt_frame {
+    frame_type_t type;
+    int line;
+
+    union {
+            const pt_fn_details_t *details;
+            const void *frame_sig;
```

Comment:
Can frame_sig be a lua_Cfunction instead of void*?

Reply:
Yes, switched to `lua_CFunction`.